data/blog/perfectly-fitting-text-to-container-in-react.md

```diff
@@ -0,0 +1,295 @@
+
+### Approach 4: JavaScript
+
+There is prior art for this. jQuery plugins like [textFix](https://github.com/STRML/textFit) solve this problem by iteratively updating the font size of the element until it fits perfectly. The downside is, you're using JavaScript, in a loop, to measure and set font size. Performance-conscious developers hate this. The upside is, you're using regular DOM text elements! The scaling will be correct (modern browsers support sub-pixel scaling just fine), the text is selectable, announce-able, the whole deal. Spoiler: this is what we went with.
+
+### Bonus Half Approch: `canvas`
+
+There's a way (I hear) to render text to a canvas and use `measureText` to get the dimensions, but I don't have a lot of `canvas` experience, I don't know what the A11y implications are, I'm loath to jump out of React too much, etc.
+
+## Attempt 1: Driving UI Through React State
+
+The [first version of the component](https://github.com/getsentry/sentry/pull/76209/commits/6a982d0167a0f290eff32bd46524aa72161184a9) stored everything in React state and drove UI updates through state updates. Here's a pseudocode version:
+
+```tsx
+function AutoSizedText({ children, minFontSize, maxFontSize, calculationCountLimit }: Props) {
+  // Set up state variables for:
+  // 1. Parent element height and width
+  // 2. The current font size
+  // 3. The most recent font size bounds
+  // 4. Calculation count
+
+  // Set up refs for:
+  // 1. Wrapper of `ParentElement`
+  // 2. Wrapper of `ChildElement`
+
+  useResizeObserver({
+    ref: parentRef,
+    onResize: () => {
+      // Reset the font size and its bounds
+      // Store the parent width and height in React state
+    };
+  })
+
+  useLayoutEffect(() => {
+    // Get parent and children dimensions
+    // Check the difference in parent and child dimensions
+    // Run the resizing algorithm
+  })
+
+  return (
+    <ParentElement>
+      <ChildElement>
+        {children}
+      </ChildElement>
+    </ParentElement>
+  )
+}
+```
+
+- Almost everything is driven via React state. When `useResizeObserver` runs, it updates the state. The state update triggers a re-render. The re-render triggers `useLayoutEffect`. The layout effect updates the size. `useResizeObserver` runs
+- Only HTML elements are in `ref`s
+- The `useLayoutEffect` hook has a dependency on literally every piece of React state, so it runs on pretty much every render
+- Each iteration of the resize algorithm updates the state, which triggers another run of the algorithm (more on the algorithm later)
+- The component uses its a `ParentElement` to wrap the child, for additional control
+
+This, in my opinion, is a natural approach. It uses basic React primitives, generously assigns state, and uses the natural render lifecycle. The only interesting thing, in my opinion, is the use of `useLayoutEffect` over `useEffect`. `useLayoutEffect` blocks browser paint. More on this later, too.
```

Comment:
```suggestion
This, in my opinion, is a natural approach. It uses basic React primitives, generously assigns state, and uses the natural render lifecycle. The only interesting thing, in my opinion, is the use of `useLayoutEffect` over `useEffect`. `useLayoutEffect` blocks browser paint. More on this later, too.
```

True, and I agree, but it is also one that is unnecessary and most often leads to performance issues. The question is why do you even need react for this?

Reply:
> The question is why do you even need react for this

Hm, what do you mean?

Reply:
Ah yes I see what you mean! It doesn't need React state at all (I mean, clearly, that got thrown out). What I'm trying to express is that it's natural to _start_ with using React state, since React's whole rendering mechanism is to inflate a DOM from a state, and then move away from using state as a future iteration